actionview/lib/action_view/renderer/partial_renderer.rb

```diff
@@ -456,12 +456,17 @@ def collection_without_template
           index = partial_iteration.index
           path, as, counter, iteration = collection_data[index]
 
+          if layout = @options[:layout]
+            layout = (cache[layout] ||= find_template(layout, keys + [as, counter, iteration]))
```

Comment:
I think that there is a problem. `collection_without_template` is used for heterogeneous collection. So `[as, counter, iteration]` is not always a same even if cache key is a same (`cache[layout] ||= ...`). Isn't it?

Reply:
I meant if using `layout` as a cache key (`cache[layout] ||= ...`), `GoodCustomer` will get wrong layout cache for `Customer`.

```diff
--- a/actionview/lib/action_view/renderer/partial_renderer.rb
+++ b/actionview/lib/action_view/renderer/partial_renderer.rb
@@ -456,8 +456,14 @@ def collection_without_template
           index = partial_iteration.index
           path, as, counter, iteration = collection_data[index]
 
+          require 'pp'
           if layout = @options[:layout]
             layout = (cache[layout] ||= find_template(layout, keys + [as, counter, iteration]))
+            pp({
+              key: @options[:layout],
+              original: [as, counter, iteration],
+              cached: layout.locals - keys.map(&:to_s)
+            })
           end
 
           locals[as]        = object
@@ -465,6 +471,11 @@ def collection_without_template
           locals[iteration] = partial_iteration
 
           template = (cache[path] ||= find_template(path, keys + [as, counter, iteration]))
+          pp({
+            key: path,
+            original: [as, counter, iteration],
+            cached: template.locals - keys.map(&:to_s)
+          })
           content = template.render(view, locals)
           content = layout.render(view, locals) { content } if layout
           partial_iteration.iterate!
```

```
% be ruby -w -Itest test/template/render_test.rb -n "CachedViewRenderTest#test_render_partial_with_layout_and_heterogeneous_collection"
Run options: -n CachedViewRenderTest#test_render_partial_with_layout_and_heterogeneous_collection --seed 25120

# Running:

{:key=>"test/b_layout_for_partial",
 :original=>[:customer, :customer_counter, :customer_iteration],
 :cached=>["customer", "customer_counter", "customer_iteration"]}
{:key=>"customers/customer",
 :original=>[:customer, :customer_counter, :customer_iteration],
 :cached=>["customer", "customer_counter", "customer_iteration"]}
{:key=>"test/b_layout_for_partial",
 :original=>[:good_customer, :good_customer_counter, :good_customer_iteration],
 :cached=>["customer", "customer_counter", "customer_iteration"]}
{:key=>"good_customers/good_customer",
 :original=>[:good_customer, :good_customer_counter, :good_customer_iteration],
 :cached=>
  ["good_customer", "good_customer_counter", "good_customer_iteration"]}
.

Finished in 0.098434s, 10.1590 runs/s, 30.4771 assertions/s.
1 runs, 3 assertions, 0 failures, 0 errors, 0 skips
```